wire/msgmixcm.go

```diff
@@ -0,0 +1,220 @@
+	Mix       *MsgTx
+	SeenDCs   []chainhash.Hash
+}
+
+// BtcDecode decodes r using the Decred protocol encoding into the receiver.
+// This is part of the Message interface implementation.
+func (msg *MsgMixCM) BtcDecode(r io.Reader, pver uint32) error {
+	const op = "MsgMixCM.BtcDecode"
+	if pver < MixVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	err := readElements(r, &msg.Signature, &msg.Identity, &msg.SessionID,
+		&msg.Expiry, &msg.Run)
+	if err != nil {
+		return err
+	}
+
+	if msg.Mix == nil {
+		msg.Mix = NewMsgTx()
+	}
+	err = msg.Mix.BtcDecode(r, pver)
+	if err != nil {
+		return err
+	}
+
+	count, err := ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+	if count > MaxPrevMixMsgs {
+		msg := fmt.Sprintf("too many previous referenced messages [%v]", count)
+		return messageError(op, ErrTooManyPrevMixMsgs, msg)
+	}
+
+	seen := make([]chainhash.Hash, count)
+	for i := range seen {
+		err := readElement(r, &seen[i])
+		if err != nil {
+			return err
+		}
+	}
+	msg.SeenDCs = seen
+
+	return nil
+}
+
+// BtcEncode encodes the receiver to w using the Decred protocol encoding.
+// This is part of the Message interface implementation.
+func (msg *MsgMixCM) BtcEncode(w io.Writer, pver uint32) error {
+	const op = "MsgMixCM.BtcEncode"
+	if pver < MixVersion {
+		msg := fmt.Sprintf("%s message invalid for protocol version %d",
+			msg.Command(), pver)
+		return messageError(op, ErrMsgInvalidForPVer, msg)
+	}
+
+	err := writeElement(w, &msg.Signature)
+	if err != nil {
+		return err
+	}
+
+	err = msg.writeMessageNoSignature(op, w, pver)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeMessageNoSignature serializes all elements of the message except for
+// the signature.  This allows code reuse between message serialization, and
+// signing and verifying these message contents.
+func (msg *MsgMixCM) writeMessageNoSignature(op string, w io.Writer, pver uint32) error {
+	err := writeElements(w, &msg.Identity, &msg.SessionID, msg.Expiry,
+		msg.Run)
+	if err != nil {
+		return err
+	}
+
+	if msg.Mix == nil {
+		msg := "nil mix transaction"
+		return messageError(op, ErrInvalidMsg, msg)
+	}
+	err = msg.Mix.BtcEncode(w, pver)
+	if err != nil {
+		return err
+	}
+
+	count := len(msg.SeenDCs)
+	if count > MaxPrevMixMsgs {
+		msg := fmt.Sprintf("too many previous referenced messages [%v]", count)
+		return messageError(op, ErrTooManyPrevMixMsgs, msg)
+	}
+
+	err = WriteVarInt(w, pver, uint64(count))
+	if err != nil {
+		return err
+	}
+	for i := range msg.SeenDCs {
+		err = writeElement(w, &msg.SeenDCs[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// WriteSigned writes a tag identifying the message data, followed by all
+// message fields excluding the signature.  This is the data committed to when
+// the message is signed.
+func (msg *MsgMixCM) WriteSigned(w io.Writer) error {
```

Comment:
I'm not quite sure that I understand the problem.  These methods hardcode the protocol version used by this serialization.  If a negotiated protocol version increases in the future, and these messages change their serialization, this function will remain consistent in what it writes to the hasher (w).

Reply:
Well, as I originally said, it's not super clear at this stage of the review, but it looks like this is supposed to be a stable serialization for the purposes of a commitment.

What I'm suggesting is that, if it is indeed something that has to be a long-term stable serialization, it should not be tied to the **wire** protocol serialization.  This is the mistake that the blocks and transactions made as it means you can't change the wire encoding to make it more efficient without also breaking consensus since it would change the hash even though that are very much two different things.

If, on the other hand, this is _only_ part of the protocol, then it isn't a problem.

What is making it unclear to me is the fact it is writing a var string of `CmdMixCM+"-sig"` followed by the rest of the confirm mix message which appears as though it's a long-term commitment.

Reply:
It's not long term at all. We are only verifying messages received on the wire originate from the claimed identity, before they will be considered to be accepted by the mixpool.  We don't store these commitments anywhere.